# GROUP 4: SONG RECOMMENDATION PROJECT

# Introduction
<b>Recommendation systems are probably the most trending data science application today. They can be used to predict users rating or preference for any particular item. All of the major tech companies are using recommendation system in some form or the other. Amazon is using it to suggest "frequently bought together" or "Customers who viewed this item also viewed". YouTube is using it to create an auto playlist based on your preferences. Infact, for companies like Netflix and Spotify, the entire business model and its success revolves around how good their recommandation system is. What's more, Netflix offered a million dollar prize competition in year 2009 to improve its system prediction by 10%.</b>

# Problem statement
<b>Build a song recommendation system which can recommend songs to listeners based on information on users behaviors, activities or preferences and predicting what users will like based on their similarity to other users”. </b>

</b>Data Source: https://www.kaggle.com/c/msdchallenge/data </b>

<b>Loading the data </b>

In [5]:
# File location and type
file_song_data = "/FileStore/tables/song_data.csv"
file_triplets_data ="/FileStore/tables/triplets_10000.txt"

<b>Uploading the data to the dataframe</b>

In [7]:
#Uploading Song CSV file with tab delimited
song_df = spark.read.csv(file_song_data,
                            inferSchema ='true',
                            header = 'true',
                            sep=',')


display(song_df)

song_id title release artist_name year SOQMMHC12AB0180CB8 Silent Night Monster Ballads X-Mas Faster Pussy cat 2003 SOVFVAK12A8C1350D9 Tanssi vaan Karkuteillä Karkkiautomaatti 1995 SOGTUKN12AB017F4F1 No One Could Ever Butter Hudson Mohawke 2006 SOBNYVR12A8C13558C Si Vos Querés De Culo Yerba Brava 2003 SOHSBXH12A8C13B0DF Tangle Of Aspens Rene Ablaze Presents Winter Sessions Der Mystic 0 SOZVAPQ12A8C13B63C "Symphony No. 1 G minor ""Sinfonie Serieuse""/Allegro con energia" Berwald: Symphonies Nos. 1/2/3/4 David Montgomery 0 SOQVRHI12A6D4FB2D7 We Have Got Love Strictly The Best Vol. 34 Sasha / Turbulence 0 SOEYRFT12AB018936C 2 Da Beat Ch'yall Da Bomb Kris Kross 1993 SOPMIYT12A6D4F851E Goodbye Danny Boy Joseph Locke 0 SOJCFMH12A8C13B0C2 Mama_ mama can't you see ? March to cadence with the US marines The Sun Harbor's Chorus-Documentary Recordings 0 SOYGNWH12AB018191E L'antarctique Des cobras des tarentules 3 Gars Su'l Sofa 2007 SOLJTLX12AB01890ED El hijo del pueblo 32 Grandes Éxitos CD 2 Jorge Negrete 1997 SOQQESG12A58A7AA28 Cold Beer feat. Prince Metropolitan International Hardcore Superstar Danny Diablo 0 SOMPVQB12A8C1379BB Pilots The Loyal Tiger Lou 2005 SOGPCJI12A8C13CCA0 N Gana Afropea 3 - Telling Stories To The Sea Waldemar Bastos 0 SOSDCFG12AB0184647 006 Lena 20 År Lena Philipsson 1998 SOBARPM12A8C133DFF (Looking For) The Heart Of Saturday Cover Girl Shawn Colvin 1994 SOKOVRQ12A8C142811 Ethos of Coercion Descend Into Depravity Dying Fetus 2009 SOIMMJJ12AF72AD643 Rock-N-Rule I'm Only A Man (Bonus Track Version) Emery 2007 SOVMBTP12A8C13A8F6 La bola extra La bola extra Los Ronaldos 0 SOOUESZ12AB0189AFD I Made It Over Let's Celebrate (He Is Risen) Rev. Timothy Wright 0 SOAGMGG12A6D4F9099 Debussy : 12 Etudes : VI Pour les huit doigts Debussy : 12 Etudes_ Images Sets 1 & 2 Pierre-Laurent Aimard 0 SOGFWVT12A8C137C64 Nervous Let No One Live Rent Free In Your Head Nicolette 1996 SOKLPMH12AB01861FA In The Journey In The Journey Martin Sexton 2001 SOEPAIN12A8C1396A7 Fuckin Ethic People (999) Crazeë Musick Craze 0 SOQMZZI12AB01850D4 Tu Vida Con La Mía Hoy Quiero Soñar Christian Castro 0 SOEEHEY12CF5F88FB4 I'm Ready Honkin' On Bobo Aerosmith 2004 SOWUMAZ12A67ADE769 Take As Needed Nerve Damage Skinlab 2002 SOJJAZT12AC4687219 Autumn In The Mind (Acoustic Version) Lo Fi_ Acoustic Excursions Mike Badger and Friends 0 SOMLHJQ12AC4685B45 Tequila Tequila il mambo del giubileo (12 Inc) Piotta 0 SOYNOIK12A6D4F7F3F All For A Purpose (Speak Life Album Version) Speak Life Sev Statik 0 SOPUJOM12AB017FEEB C On C On The Advent 0 SOGNNYL12A6D4F910B Raspberry Beret (LP Version) The Hits/The B-Sides Prince & The Revolution 0 SOOLRHW12A8C142643 All of the same blood Violent revolution Kreator 2001 SOMKNZS12A8C13DEBC One Little Too Little Hall Of Fame Hall Of Fame 1996 SOYKVON12A8C14097E Cockleshell Heroes Classic Years Frank Chacksfield 2007 SOEXAWC12AB01817E6 Trancesequence Voyage Voyage 0 SOKQCDA12A6D4FA556 Wonderful Stash TRANCE VISIONARY Wishbone Ash 1998 SOFWVLR12A8C1328DC La filo Le meilleur Malavoi 0 SOLRKBF12A8C13C88A Throw It Away Easy Listening Cobra Verde 2003 SOBDLRM12A8C13A0AC Before He Kissed Me Lisa Brokop Lisa Brokop 0 SOMXBRA12AB017C656 Blessed Rejoice The Emotions 1977 SODJAOH12A58A77B23 Goa Amsterdam B.P.M. - Bionic Pulse Method Vol. 3 Safi Connection 0 SOCPTIN12A8C14265F Disaster [Demo Version] Fragile Future Hawthorne Heights 2008 SODVOFJ12AB0181EE6 Night And Day The Best Of The Maytals 1994 SODDEQU12AAF3B2FC8 Passione Alla Fine Della Notte Neffa 2007 SONYUEW12AB018B373 My Everything [Screwed] (feat. Trae The Truth) [Screwed] Tha Flood - Swishahouse Mix Lil O 0 SOWWQXA12A81C1F9CA Scream Devil's Playground Billy Idol 2005 SONNHYN12AB018538B The Man Delusion The Man Delusion The Vichy Government 0 SOCWWYB12A6D4F9AF4 He Is Not Silent (Out Of The Grey Album Version) The Early Years Out Of The Grey 0 SOIRTNX12AB018B41D Pepe El Mañoso Salsa Cubana Vol.6 Orquesta Sonara La Habana 0 SOLGCKQ12A8AE47CFA Acércate Más Éxitos Del Romántico De Siemp

<b>Defining schema for triplets data</b>

In [9]:
from pyspark.sql.types import *

# Creating schema
schema = StructType([StructField('user_id', StringType()),
                      StructField('songid', StringType()),
                      StructField('play_count', IntegerType())])

#Uploading Triplets file with tab delimited
tri_df = spark.read.csv(file_triplets_data,
                          schema= schema,
                            sep='\t')


display(tri_df)

user_id songid play_count b80344d063b5ccb3212f76538f3d9e43d87dca9e SOAKIMP12A8C130995 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOBBMDR12A8C13253B 2 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOBXHDL12A81C204C0 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOBYHAJ12A6701BF1D 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SODACBL12A8C13C273 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SODDNQT12A6D4F5F7E 5 b80344d063b5ccb3212f76538f3d9e43d87dca9e SODXRTY12AB0180F3B 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOFGUAY12AB017B0A8 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOFRQTD12A81C233C0 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOHQWYZ12A6D4FA701 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOIYTOA12A6D4F9A23 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOIZAZL12A6701C53B 5 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOJNNUA12A8AE48C7A 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOJPFQG12A58A7833A 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOKRIMP12A6D4F5DA3 5 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOLLGNU12AF72A4D4F 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOMGIYR12AB0187973 6 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOMLMKI12A81C204BC 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOMSQJY12A8C138539 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SONSAEZ12A8C138D7A 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOOKGRB12A8C13CD66 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOPCVQE12AC468AF36 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOQIVUD12AB01821D2 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOQJLDY12AAF3B456D 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOQLCKR12A81C22440 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SORPMYJ12AF729EB90 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SORQHCG12A58A7EEBA 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SORUFVF12AB018230B 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SORWLTW12A670208FA 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SORZASF12A6D4F8CFA 2 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOSYBEV12AB0182933 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOTFATN12A6D4FA74D 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOTLVCL12AB0182D22 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOTRSFZ12A8C142BF6 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOUKXIN12A8C133C7F 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOVHRGF12A8C13852F 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOVQEYZ12A8C1379D8 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOVYIYI12A8C138D88 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOWQLXP12AF72A08A2 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOWSPUS12AC468BEE3 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOXMIUS12A8C13CD59 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOXRXDG12A8C131DE5 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOXZQDE12A8C135833 1 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOYHEPA12A8C13097F 8 b80344d063b5ccb3212f76538f3d9e43d87dca9e SOZOBWN12A8C130999 1 85c1f87fea955d09b4bec2e36aee110927aedf9a SODJTHN12AF72A8FCD 2 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOBDRND12A8C13FD08 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOCHBAJ12AAF3B3A4F 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOCZTMT12AF72A078E 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOHRQZQ12A6D4F81D2 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOJGMYY12AB01809BE 2 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOQFEDG12AB018DD24 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOVRZIX12AAF3B2A32 1 bd4c6e843f00bd476847fb75c47b4fb430a06856 SOZMJFG12AB017BDAF 1 8937134734f869debcab8f23d77465b4caaa85df SOJQOIK12AF72A0AAF 5 8937134734f869debcab8f23d77465b4caaa85df SOVHVBJ12A58A7AB0A 5 969cc6fb74e076a68e36a04409cb9d3765757508 SOABRAB12A6D4F7AAF 2 969cc6fb74e076a68e36a04409cb9d3765757508 SOAOQFD12A6D4FAAA9 1 969cc6fb74e076a68e36a04409cb9d3765757508 SOBFPJC12A58A7D1AB 9 969cc6fb74e076a68e36a04409cb9d3765757508 SOBZZDU12A6310D8A3 2 969cc6fb74e076a68e36a04409cb9d3765757508 SODCNJX12A6D4F93CB 2 969cc6fb74e076a68e36a04409cb9d3765757508 SOERVXG12A6D4F5583 1 969cc6fb74e076a68e36a04409cb9d3765757508 SOFXZAW12A67ADC90E 1 969cc6fb74e076a68e36a04409cb9d3765757508 SOGFUFC12A8C13F1E5 6 969cc6

# Implicit VS Explicit data:

<b>Explicit data is the data where we have user rating associated with a song or a movie on a fixed scale. For instance: 1 to 5 ratings in the Netflix dataset. From such rating, we can interpret how much a user likes or dislikes a movie, but it is hard to get such data because generally users do not care to rate every movie they see.<br><br>
Implicit data is the type of data we are using for song recommendation. The data is gathered from the user behavior, with no explicit rating associated with it. It could be how many times a user played a song or watched a movie, how long they have a spent reading a particular article etc. The advantage here is we have a lot of such data but it is usually very noisy and unreliable.<br><br>
When a user rates a movie 1 on scale of 5 that means that he did not like the movie. But with play count of a song it we can’t make any implicit assumption that the user loved the song or hated the song or somewhere-in-between. Also, if they did not play a song does not necessarily mean that they do not like the song.
Therefore, we focus on what we know about the users behavior and the confidence we have in whether or not they like any given item. For instance: we can have a higher confidence on a song if the user played it 100 times against a song which he played on 1 time.</b>

<b>Joining the two dataframes to form MSD Data having the playcounts from tri_df</b>

In [12]:
MSD = tri_df.join(song_df, tri_df.songid == song_df.song_id,how='left') 
MSD.show(5)

+--------------------+------------------+----------+------------------+-----+------------------+-----------+----+
 user_id| songid|play_count| song_id|title| release|artist_name|year|
+--------------------+------------------+----------+------------------+-----+------------------+-----------+----+
79f93851e840f9d1f...|SOATHTW12A58A7EDB5| 1|SOATHTW12A58A7EDB5| Mutt|Enema Of The State| Blink-182|1998|
043d81932e75d5749...|SOATHTW12A58A7EDB5| 5|SOATHTW12A58A7EDB5| Mutt|Enema Of The State| Blink-182|1998|
ebacfcb5fa29a601f...|SOATHTW12A58A7EDB5| 1|SOATHTW12A58A7EDB5| Mutt|Enema Of The State| Blink-182|1998|
417c73dd95669d191...|SOATHTW12A58A7EDB5| 1|SOATHTW12A58A7EDB5| Mutt|Enema Of The State| Blink-182|1998|
52ab33fbb2fa3aeb2...|SOATHTW12A58A7EDB5| 1|SOATHTW12A58A7EDB5| Mutt|Enema Of The State| Blink-182|1998|
+--------------------+------------------+----------+------------------+-----+------------------+-----------+----+
only showing top 5 rows

<b> Removing the redundant column </b>

In [14]:
MSD = MSD['user_id','song_id','play_count','title','release','artist_name','year']
MSD.show(5)

+--------------------+------------------+----------+-----+------------------+-----------+----+
 user_id| song_id|play_count|title| release|artist_name|year|
+--------------------+------------------+----------+-----+------------------+-----------+----+
79f93851e840f9d1f...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182|1998|
043d81932e75d5749...|SOATHTW12A58A7EDB5| 5| Mutt|Enema Of The State| Blink-182|1998|
ebacfcb5fa29a601f...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182|1998|
417c73dd95669d191...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182|1998|
52ab33fbb2fa3aeb2...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182|1998|
+--------------------+------------------+----------+-----+------------------+-----------+----+
only showing top 5 rows

<b>Changing the column name year to release_year</b>

In [16]:
MSD = MSD.withColumnRenamed('year', 'release_year')
MSD.show(5)


+--------------------+------------------+----------+-----+------------------+-----------+------------+
 user_id| song_id|play_count|title| release|artist_name|release_year|
+--------------------+------------------+----------+-----+------------------+-----------+------------+
79f93851e840f9d1f...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182| 1998|
043d81932e75d5749...|SOATHTW12A58A7EDB5| 5| Mutt|Enema Of The State| Blink-182| 1998|
ebacfcb5fa29a601f...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182| 1998|
417c73dd95669d191...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182| 1998|
52ab33fbb2fa3aeb2...|SOATHTW12A58A7EDB5| 1| Mutt|Enema Of The State| Blink-182| 1998|
+--------------------+------------------+----------+-----+------------------+-----------+------------+
only showing top 5 rows

<b>Identifying total number of distinct songs and users</b>

In [18]:
# Number of rows 
print(MSD.count())
print(MSD.select("user_id").distinct().count())
print(MSD.select("song_id").distinct().count())

2086946
76353
10000

#Data Exploration

In [20]:
# Create a view or table

temp_table_name = "user_playlist"

MSD.createOrReplaceTempView(temp_table_name)

In [21]:
%sql
select * from user_playlist

user_id song_id play_count title release artist_name release_year 79f93851e840f9d1faeba586ee18b30fdb0008b6 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 043d81932e75d5749ed5758d6420506e7bc457a5 SOATHTW12A58A7EDB5 5 Mutt Enema Of The State Blink-182 1998 ebacfcb5fa29a601f596b2d1076d7973177737e1 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 417c73dd95669d1919c869ef20fd2d0f7a31403d SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 52ab33fbb2fa3aeb2a261734603061e288a2b253 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 13ce57b3a25ef63fa614335fd838e8024c42ec17 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 3294ef9047ac6bf73c8c6fb14a3096ca05a67cb6 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 8b871cbd0f9c62dc3dd349b86d855102515070de SOATHTW12A58A7EDB5 2 Mutt Enema Of The State Blink-182 1998 dd88cd67ebe00d6a81a040df0d3be9390927d399 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 b1f0e90e73e6a786bd2b3b17f6e917a777e6e19b SOATHTW12A58A7EDB5 10 Mutt Enema Of The State Blink-182 1998 e9b70166b2734172ad0a08983c54c4a0858dc862 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 d0f0134594ad065e29e9a236da4c38e7ab7bd3e8 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 fef45cbcef5bf7571d053652c9c46085a3ee38a4 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 17f988b9693ba5f0e46817f0c7fdec9370b46e19 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 139597326c92d066c1cd7cda88a6847409cda089 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 646257b8127326b462aae190c8c75dd3cce49251 SOATHTW12A58A7EDB5 2 Mutt Enema Of The State Blink-182 1998 da7a82a689f25f5de92310c1e2a9d34c44c89c93 SOATHTW12A58A7EDB5 2 Mutt Enema Of The State Blink-182 1998 ef55d17b3d3b6a9dd035efd8d107f7646c607812 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 1db12101f3e1a943c68c15d07739a71c9cc82ecc SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 cb194b8f02aa34b3c5b2e7857c988d4ce31ee5b7 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 f2c1bff79c31aa6598963ae9c1c91fa7d2746094 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 90b540cbb748698bed8eed0803b8b982b742317a SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 a1ad2731314ed3843e7b256365d5ef674aabd3b2 SOATHTW12A58A7EDB5 7 Mutt Enema Of The State Blink-182 1998 1ff5a30108e1a2dd8e54deac030c262b94249802 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 deb0f5d1b5d7c6b3d9daa08b4cc8a36ff6fe3b69 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 7599d575390d85814df47f7294d8f39cac17568a SOATHTW12A58A7EDB5 5 Mutt Enema Of The State Blink-182 1998 b776f599b68252d8d55e361ee025b863c5a63f45 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 89d8b7d2e1a26fc9fca3afd385a2c41067e65626 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 de718eb06d901cfdd9e7e20dc9ed6b9ac6d6e1ed SOATHTW12A58A7EDB5 14 Mutt Enema Of The State Blink-182 1998 04b64d358586437f28ec11cb0782e273dc689f49 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 7f8e2c875815d75bb4ff34afabcb60b305e54870 SOATHTW12A58A7EDB5 4 Mutt Enema Of The State Blink-182 1998 fff543db7918cb8f4f56f7470903eb2f1d5a6dd8 SOATHTW12A58A7EDB5 3 Mutt Enema Of The State Blink-182 1998 296b8fffb87fb68a69f09b5c344ecffcd91100ee SOATHTW12A58A7EDB5 2 Mutt Enema Of The State Blink-182 1998 8864d6bd98b8cfeada01bb63a36c18bb9166f378 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 691d3d643a4327aab5d84e2e0d12cb0c3369158f SOATHTW12A58A7EDB5 6 Mutt Enema Of The State Blink-182 1998 9dc4a128934f3bd9353ef75b9efc260b23502415 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 94024e65cc32f00d688dba893832ed198e28994d SOATHTW12A58A7EDB5 5 Mutt Enema Of The State Blink-182 1998 52b331848be4d6c305448e9758a66baa00a84f20 SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-182 1998 f4717057b016090e138e25098b22b8eadd5e98ab SOATHTW12A58A7EDB5 1 Mutt Enema Of The State Blink-18

<b>Most played 10 Songs and there Artists</b>

In [23]:
%sql
select artist_name, title, sum(play_count) as number_of_total_play from user_playlist group by title,artist_name order by sum(play_count) desc limit 10;

artist_name title number_of_total_play Dwight Yoakam You're The One 54136 Björk Undo 49253 Kings Of Leon Revelry 41418 Barry Tuckwell/Academy of St Martin-in-the-Fields/Sir Neville Marriner Horn Concerto No. 4 in E flat K495: II. Romance (Andante cantabile) 31153 Harmonia Sehr kosmisch 31036 Florence + The Machine Dog Days Are Over (Radio Edit) 26663 Kings Of Leon Use Somebody 22140 OneRepublic Secrets 22100 Five Iron Frenzy Canada 21019 Tub Ring Invalid 19645

<b>Top 10 listeners</b>

In [25]:
%sql
select user_id, sum(play_count) as number_of_total_play from user_playlist group by user_id order by sum(play_count) desc limit 10;

user_id number_of_total_play 4be305e02f4e72dad1b8ac78e630403543bab994 4884 6d625c6557df84b60d90426c0116138b617b9449 3548 972cce803aa7beceaa7d0039e4c7c0ff097e4d55 3399 0b19fe0fad7ca85693846f7dad047c449784647e 3059 d13609d62db6df876d3cc388225478618bb7b912 2728 283882c3d18ff2ad0e17124002ec02b847d06e9a 2322 083a2a59603a605275107c00812a811526c2a0af 2289 2231cb435771a1a621ec44e95cdd28b81fad3288 2205 6a944bfe30ae8d6b873139e8305ae131f1607d5f 2170 9c859962257112ad523f1d3c121d35191daa6d2b 2124

<b>Distribution of Play count for all songs</b>

In [27]:
%sql
select play_count, count(*) as count from user_playlist group by play_count order by play_count

play_count count 1 1188874 2 327381 3 149521 4 86400 5 96247 6 47439 7 32289 8 23514 9 17436 10 19566 11 13010 12 10248 13 8323 14 6699 15 7158 16 5535 17 4640 18 3880 19 3337 20 3433 21 2726 22 2422 23 2068 24 1913 25 1848 26 1554 27 1344 28 1230 29 1107 30 1040 31 982 32 879 33 761 34 689 35 725 36 666 37 537 38 510 39 498 40 505 41 423 42 382 43 338 44 314 45 322 46 306 47 263 48 262 49 259 50 244 51 243 52 210 53 183 54 189 55 194 56 158 57 154 58 145 59 131 60 155 61 127 62 114 63 94 64 98 65 100 66 89 67 92 68 77 69 81 70 77 71 81 72 70 73 63 74 65 75 68 76 56 77 60 78 46 79 50 80 60 81 43 82 44 83 49 84 45 85 46 86 41 87 34 88 33 89 39 90 28 91 32 92 37 93 32 94 37 95 36 96 27 97 31 98 15 99 22 100 21 101 29 102 16 103 24 104 23 105 19 106 11 107 21 108 11 109 13 110 14 111 16 112 14 113 14 114 13 115 20 116 10 117 8 118 13 119 14 120 11 121 12 122 10 123 18 124 6 125 13 126 14 127 11 128 7 129 7 130 10 131 5 132 9 133 6 134 5 135 4 136 6 137 9 138 9 139 8 140 9 141 7 142 3 143 4 144 13 145 5 146 7 147 6 148 7 149 6 150 11 151 3 152 4 153 5 154 4 155 10 156 9 157 5 158 3 159 3 160 4 161 3 162 2 163 6 164 7 165 3 166 3 167 4 168 4 169 2 170 3 171 4 172 3 173 3 174 4 175 8 176 1 177 3 178 5 179 4 180 4 181 2 182 5 183 1 184 5 185 1 186 5 187 2 188 5 189 2 190 1 191 1 192 2 193 2 194 3 195 1 196 1 197 2 198 3 199 1 200 2 201 1 202 2 204 2 205 4 206 4 207 3 208 3 211 1 215 2 217 1 218 2 219 1 220 1 221 2 224 5 225 3 226 5 228 1 229 1 232 2 234 1 235 1 236 2 239 1 243 1 246 1 247 2 250 2 251 1 252 3 254 1 255 1 259 1 260 1 261 2 265 2 267 3 270 1 271 1 272 1 274 1 277 1 279 1 283 1 284 1 286 1 292 2 293 1 295 1 297 1 302 1 305 1 307 1 310 2 311 2 317 1 319 1 322 1 328 1 330 1 333 4 338 1 353 3 356 1 359 1 364 1 372 2 375 1 383 1 392 2 401 1 405 1 413 1 415 1 416 1 419 1 422 1 431 1 436 1 446 1 453 1 455 1 488 1 500 2 553 1 585 1 605 1 649 1 664 3 676 1 683 1 796 1 879 1 920 1 2213 1

<b>Top played songs based on Artists</b>

In [29]:
%sql
select artist_name, sum(play_count) as number_of_total_play from user_playlist group by artist_name order by sum(play_count) desc limit 10;

artist_name number_of_total_play Kings Of Leon 86031 Coldplay 78540 Florence + The Machine 60066 Dwight Yoakam 54136 Björk 53814 The Black Keys 52220 Muse 52136 Justin Bieber 50376 Jack Johnson 48487 Eminem 41754

<b>Distribution of distinct songs listened by each users</b>

In [31]:
%sql
select Number_of_songs,count(*) from (
select user_id,count(distinct song_id) as Number_of_songs from user_playlist group by user_id order by Number_of_songs) group by Number_of_songs order by Number_of_songs;

Number_of_songs count(1) 1 875 2 1226 3 1585 4 1952 5 2410 6 2753 7 3070 8 3410 9 3454 10 3600 11 3242 12 2944 13 2669 14 2349 15 2237 16 2016 17 1832 18 1703 19 1602 20 1444 21 1323 22 1237 23 1172 24 1078 25 983 26 1051 27 961 28 906 29 878 30 851 31 747 32 675 33 739 34 640 35 639 36 547 37 534 38 489 39 461 40 472 41 480 42 425 43 459 44 403 45 387 46 358 47 380 48 342 49 356 50 307 51 316 52 300 53 294 54 244 55 261 56 256 57 270 58 227 59 234 60 207 61 220 62 233 63 171 64 181 65 192 66 192 67 162 68 153 69 138 70 163 71 160 72 138 73 141 74 130 75 132 76 116 77 117 78 110 79 137 80 114 81 110 82 104 83 99 84 99 85 106 86 97 87 73 88 75 89 72 90 77 91 63 92 77 93 78 94 71 95 60 96 68 97 63 98 64 99 69 100 57 101 56 102 66 103 60 104 49 105 44 106 54 107 36 108 36 109 43 110 37 111 45 112 41 113 49 114 28 115 39 116 22 117 47 118 35 119 39 120 41 121 34 122 29 123 45 124 30 125 24 126 20 127 32 128 19 129 35 130 27 131 27 132 24 133 20 134 19 135 22 136 18 137 20 138 24 139 19 140 23 141 29 142 20 143 21 144 10 145 14 146 10 147 17 148 17 149 18 150 21 151 15 152 16 153 12 154 12 155 7 156 14 157 14 158 18 159 14 160 12 161 12 162 19 163 14 164 18 165 8 166 9 167 16 168 10 169 6 170 14 171 9 172 13 173 12 174 12 175 6 176 10 177 10 178 8 179 8 180 4 181 6 182 5 183 5 184 9 185 7 186 8 187 5 188 10 189 12 190 10 191 5 192 5 193 9 194 8 195 8 196 9 197 9 198 9 199 6 200 6 201 8 202 2 203 1 204 4 205 4 206 4 207 6 208 8 209 8 210 3 211 1 212 2 213 5 214 5 215 7 216 7 217 4 218 7 219 3 220 4 221 4 222 4 223 2 224 5 225 4 226 2 227 8 228 2 229 1 230 2 231 1 232 3 233 2 235 4 236 3 237 7 238 6 239 1 240 5 241 4 242 8 243 2 244 5 245 3 246 6 247 1 248 2 249 2 250 2 251 2 252 3 254 3 257 4 258 2 259 2 260 1 261 4 263 4 264 1 265 4 266 1 268 2 269 2 270 1 272 2 273 3 274 1 275 5 276 1 277 3 278 2 279 3 280 2 282 2 283 3 285 1 286 2 287 2 289 1 290 5 291 2 292 2 294 1 295 1 296 2 299 1 300 2 301 4 302 1 303 1 304 1 305 2 309 1 310 1 312 1 313 2 314 1 315 3 318 2 323 2 325 1 328 1 329 2 330 2 332 1 333 1 334 1 337 1 338 1 340 1 347 1 350 2 351 1 354 1 355 1 358 1 364 1 365 1 371 1 377 1 380 1 383 2 384 1 401 2 403 2 409 1 422 1 423 1 426 1 452 1 473 1 474 1 478 1 480 1 487 1 489 1 502 1 522 1 523 1 533 1 540 1 556 1 643 1 711 1

<b>Taking 10% of the users from MSD data</b>

In [33]:
# number of distinct user_Id
user=MSD.select("user_id").distinct()
user1,user2= user.randomSplit([0.05,0.95], seed=123)
usercount = user1.count()
print("Number of users: ", usercount)

Number of users: 3783

<b>Dataframe of distinct list of songs</b>

In [35]:
# number of distinct song_Id
songs= MSD.select("song_id").distinct()
songcount=songs.count()
print("Number of songs: ", songcount)

Number of songs: 10000

<b>Giving incrementing Ids to UserId field to convert them into integer such that it can be accepted by our model</b>

In [37]:
from pyspark.sql.functions import monotonically_increasing_id

# Creating new columns of unique integers for user_id and song_id
user_df = user1.withColumn("new_userid", monotonically_increasing_id())
user_df.show()

+--------------------+----------+
 user_id|new_userid|
+--------------------+----------+
126ef5859eb5e96f7...| 0|
1d7b9780e492c062c...| 1|
2c0815308dfd33b4b...| 2|
3d24b9ffed6a82778...| 3|
3dd54878cb47456b8...| 4|
459ab8388e7806755...| 5|
5e7105e485a04bf0b...| 6|
6bb6fa9a23505dc55...| 7|
739f32d4c2690554b...| 8|
7a3943dfa7f83e321...| 9|
7de4388c64742657d...| 10|
9619f405e777e8331...| 11|
a2c1d795852bd22c6...| 12|
a8e8fd13a2909af99...| 13|
aab99ec2a563f732e...| 14|
c06d794619168b4bf...| 15|
c2cfc654c54fddc9f...| 16|
e43304402c7407bc4...| 17|
e6bf98dccce485c26...| 18|
e7fc73d0eb0d851bc...| 19|
+--------------------+----------+
only showing top 20 rows

<b>Giving incrementing Ids to SongId field to convert them into integer such that it can be accepted by our model</b>

In [39]:
songs_df = songs.select("song_id", monotonically_increasing_id().alias('new_songId'))
songs_df.show()

+------------------+----------+
 song_id|new_songId|
+------------------+----------+
SOATHTW12A58A7EDB5| 0|
SOAZMXH12AB0186DDE| 1|
SOBAQTV12A8C142277| 2|
SOCKUUJ12A6D4FA41C| 3|
SOCUVKX12A6D4F8ED7| 4|
SODABFB12A58A81788| 5|
SODASIJ12A6D4F5D89| 6|
SODYTRD12A81C2329F| 7|
SOECOOL12AB0181A2F| 8|
SOECTGX12A6310E233| 9|
SOERLLT12AC468DAF3| 10|
SOGKEGN12AB0185355| 11|
SOGXQYC12AB0183AE5| 12|
SOHXDTJ12A81C219C2| 13|
SOICVFJ12A8AE47FF0| 14|
SOJGIUN12A6BD55B8E| 15|
SOKOVZK12A6D4F707F| 16|
SOKQHXV12AB0185B3D| 17|
SOKUAGP12A8C133B94| 18|
SOLIVXX12A6D4F7950| 19|
+------------------+----------+
only showing top 20 rows

<b>Cross joining to map all the users with all the songs such that we have entry for all the songs for each user</b>

In [41]:
#Cross Join user and Songs
crossjoin = user_df.crossJoin(songs_df)
crossjoin.show(5)

+--------------------+----------+------------------+----------+
 user_id|new_userid| song_id|new_songId|
+--------------------+----------+------------------+----------+
126ef5859eb5e96f7...| 0|SOATHTW12A58A7EDB5| 0|
126ef5859eb5e96f7...| 0|SOAZMXH12AB0186DDE| 1|
126ef5859eb5e96f7...| 0|SOBAQTV12A8C142277| 2|
126ef5859eb5e96f7...| 0|SOCKUUJ12A6D4FA41C| 3|
126ef5859eb5e96f7...| 0|SOCUVKX12A6D4F8ED7| 4|
+--------------------+----------+------------------+----------+
only showing top 5 rows

In [42]:
crossjoin.count()

Out[18]: 37830000

<b> Joining the crossjoin dataframe with entire data such that we can have play_count populated for all the user and song combination, and replacing the NA's with 0 when there is no match i.e. when user did not listen to that song
</b>

In [44]:
df = crossjoin.join(MSD, ["user_id", "song_id"], "left").fillna(0)

<b>Selecting only numeric columns that we want for Modeling</b>

In [46]:
model_df= df.select(df.new_userid.cast("int"),df.new_songId.cast("int"),df.play_count.cast("int"))

#Alternating Least Squares (ALS)
<b>ALS is an iterative optimization process in which for every run we try to arrive closer and closer to a factorized representation of our original data. Assume that our original matrix M of size U x I, where u are the number of users and i are the number of items. We want to find a way such that we can express our original matrix M into product of two matrix. One matrix of user and hidden features of dimension U * F and second matrix of items and hidden features of dimension F x I. These two matrix have weights for how each user/item relates to each hidden feature. Using gradient descent, we evalaute these two matrix such that their product approximates M as closely as possible.</b>

<b>Defining the Model</b>

In [49]:
# Set the ALS hyperparameters
from pyspark.ml.recommendation import ALS

model = ALS(userCol= "new_userid", itemCol= "new_songId", ratingCol= "play_count", rank = 10, maxIter = 10,alpha = 20, regParam = .05,  coldStartStrategy="drop", nonnegative = True, implicitPrefs = True)

<b>Dividing the data into train and test dataset</b>

In [51]:
# Split the dataframe into training and test data
(train_data, test_data) = model_df.select('new_userid','new_songId','play_count').randomSplit([0.7, 0.3], seed=12345)

# Rank Ordering Error Metric (ROEM)
<b> The ALS model from Spark ml library contains additional parameter for implicit rating called alpha which is an integer value that tells Spark how much additional song play should add to the confidence of model that the user actually likes a particular song.<br><br>
For explicit ratings, we can use RMSE to evaluate the model which makes sense we can match predictions back to a true measure of user ratings. However, in case of implicit rating we don’t have true measure of user ratings we only have the number of times user played a song and a measure of how confident the model is that they like that song and therefore we can’t use RMSE for evaluating our model. However, using test dataset, we can see if our model is giving high predictions to the songs that users have actually listened to.<br><br>
The logic is if our model is returning a high prediction for a song that the respective user actually listened to, then the model prediction make sense, especially if they’ve listened to it more than once. We can measure this using Rank Ordering Error Metric (ROEM), which checks if songs have higher number of plays have higher predictions.</b>

In [53]:
def ROEM(predictions, userCol = "new_userid", itemCol = "new_songId", ratingCol = "play_count"):
  #Creates table that can be queried
  predictions.createOrReplaceTempView("predictions")

  #Sum of total number of plays of all songs
  denominator = predictions.groupBy().sum(ratingCol).collect()[0][0]

  #Calculating rankings of songs predictions by user
  spark.sql("SELECT " + userCol + " , " + ratingCol + " , PERCENT_RANK() OVER (PARTITION BY " + userCol + " ORDER BY prediction DESC) AS rank FROM predictions").createOrReplaceTempView("rankings")

   #Multiplies the rank of each song by the number of plays and adds the products together
  numerator = spark.sql('SELECT SUM(' + ratingCol + ' * rank) FROM rankings').collect()[0][0]

  performance = numerator/denominator

  return performance

In [54]:
train_data.cache()

Out[24]: DataFrame[new_userid: int, new_songId: int, play_count: int]

<b>Fits model to fold within training data</b>

In [56]:
fitted_model = model.fit(train_data)

<b>Generates predictions using fitted_model on respective CV test data</b>

In [58]:
predictions = fitted_model.transform(test_data)

<b>Generates and prints a ROEM metric CV test data</b>

In [60]:
# Generates and prints a ROEM metric CV test data
validation_performance = ROEM(predictions)
print(validation_performance)

0.5026878341522527

# Conclusion

Our model achieved the accuracy of 50.26%. We trained the model on ~4000 users.
<br> Out of 2 two songs recommended, 1 was relevant to the user.
<br> We are able to process 37830000 records (3783 users and 10000 songs).
<br> We tried multiple cloud platforms to increase accuracy and performance of the code.(Databricks, Google Cloud Platform)
<br> The model performance can be improved if we train it on more data points.

# References
<b>https://github.com/jamenlong/ALS_expected_percent_rank_cv/blob/master/ROEM_function.py <br>
https://medium.com/radon-dev/als-implicit-collaborative-filtering-5ed653ba39fe<br>
IEEE Paper : “Collaborative Filtering for Implicit Feedback Datasets”</b><br>